File: app/src/main/java/com/boardgamegeek/ui/dialog/UpdateBuddyNicknameDialogFragment.java
package com.boardgamegeek.ui.dialog;

import android.app.Dialog;
import android.content.DialogInterface;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.StringRes;
import android.support.v4.app.DialogFragment;
import android.support.v7.app.AlertDialog;
import android.text.InputType;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager.LayoutParams;
import android.widget.CheckBox;
import android.widget.EditText;

import com.boardgamegeek.R;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;

public class UpdateBuddyNicknameDialogFragment extends DialogFragment {
	public interface UpdateBuddyNicknameDialogListener {
		void onFinishEditDialog(String newNickname, boolean shouldUpdatePlays);
	}

	private static final String KEY_TITLE_ID = "title_id";
	@StringRes private int titleResId;
	private ViewGroup root;
	private UpdateBuddyNicknameDialogListener listener;

	private Unbinder unbinder;
	@BindView(R.id.edit_nickname) EditText editText;
	@BindView(R.id.change_plays) CheckBox changePlays;
	private String nickname;

	@NonNull
	public static UpdateBuddyNicknameDialogFragment newInstance(
		@StringRes int titleResId,
		@Nullable ViewGroup root,
		UpdateBuddyNicknameDialogListener listener) {

		UpdateBuddyNicknameDialogFragment fragment = new UpdateBuddyNicknameDialogFragment();
		fragment.initialize(titleResId, root, listener);
		return fragment;
	}

	private void initialize(
		@StringRes int titleResId,
		@Nullable ViewGroup root,
		UpdateBuddyNicknameDialogListener listener) {

		this.titleResId = titleResId;
		this.root = root;
		this.listener = listener;
		setArguments(this.titleResId);
	}

	public void setArguments(@StringRes int titleResId) {
		Bundle bundle = new Bundle();
		bundle.putInt(KEY_TITLE_ID, titleResId);
		setArguments(bundle);
	}

	@Override
	@NonNull
	public Dialog onCreateDialog(Bundle savedInstanceState) {
		LayoutInflater layoutInflater = LayoutInflater.from(getActivity());
		View rootView = layoutInflater.inflate(R.layout.dialog_edit_nickname, root, false);
		unbinder = ButterKnife.bind(this, rootView);

		if (getArguments() != null) {
			titleResId = getArguments().getInt(KEY_TITLE_ID);
		}

		setAndSelectExistingText();

		AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
		if (titleResId > 0) {
			builder.setTitle(titleResId);
		}
		builder.setView(rootView)
			.setNegativeButton(R.string.cancel, null)
			.setPositiveButton(R.string.ok, new DialogInterface.OnClickListener() {
				@Override
				public void onClick(DialogInterface dialog, int which) {
					if (listener != null) {
						listener.onFinishEditDialog(editText.getText().toString().trim(), changePlays.isChecked());
					}
				}
			});

		final AlertDialog dialog = builder.create();
		int inputType = InputType.TYPE_TEXT_FLAG_CAP_WORDS;
		editText.setInputType(editText.getInputType() | inputType);
		requestFocus(dialog);
		return dialog;
	}

	@Override
	public void onDestroyView() {
		super.onDestroyView();
		if (unbinder != null) unbinder.unbind();
	}

	public void setNickname(String nickname) {
		this.nickname = nickname;
	}

	private void setAndSelectExistingText() {
		if (editText != null && !TextUtils.isEmpty(nickname)) {
			editText.setText(nickname);
			editText.setSelection(0, nickname.length());
		}
	}

	private void requestFocus(@NonNull AlertDialog dialog) {
		editText.requestFocus();
		dialog.getWindow().setSoftInputMode(LayoutParams.SOFT_INPUT_STATE_VISIBLE);
	}
}

File: app/src/main/java/com/boardgamegeek/ui/dialog/UpdateBuddyNicknameDialogFragment.kt
package com.boardgamegeek.ui.dialog

import android.annotation.SuppressLint
import android.app.Dialog
import android.os.Bundle
import android.text.InputType
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.app.AlertDialog
import androidx.fragment.app.DialogFragment
import androidx.lifecycle.ViewModelProvider
import com.boardgamegeek.R
import com.boardgamegeek.databinding.DialogEditNicknameBinding
import com.boardgamegeek.extensions.requestFocus
import com.boardgamegeek.extensions.setAndSelectExistingText
import com.boardgamegeek.ui.viewmodel.BuddyViewModel
import org.jetbrains.anko.support.v4.act

class UpdateBuddyNicknameDialogFragment : DialogFragment() {
    private var _binding: DialogEditNicknameBinding? = null
    private val binding get() = _binding!!

    private val viewModel: BuddyViewModel by lazy {
        ViewModelProvider(this).get(BuddyViewModel::class.java)
    }

    @SuppressLint("InflateParams")
    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        super.onCreateDialog(savedInstanceState)
        _binding = DialogEditNicknameBinding.inflate(LayoutInflater.from(context))

        return AlertDialog.Builder(requireContext(), R.style.Theme_bgglight_Dialog_Alert)
                .setView(binding.root)
                .setTitle(R.string.title_edit_nickname)
                .setNegativeButton(R.string.cancel, null)
                .setPositiveButton(R.string.ok) { _, _ ->
                    viewModel.updateNickName(binding.nicknameView.text.trim().toString(), binding.changePlaysCheckBox.isChecked)
                }
                .create().apply {
                    requestFocus(binding.nicknameView)
                }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        if (savedInstanceState == null) {
            binding.nicknameView.setAndSelectExistingText(arguments?.getString("NICKNAME"))
            binding.nicknameView.inputType = binding.nicknameView.inputType or InputType.TYPE_TEXT_FLAG_CAP_WORDS
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    companion object {
        @JvmStatic
        fun newInstance(nickname: String): UpdateBuddyNicknameDialogFragment {
            val fragment = UpdateBuddyNicknameDialogFragment()
            fragment.arguments = Bundle().apply {
                putString("NICKNAME", nickname)
            }
            return fragment
        }
    }
}
